File: filesystem/src/main/java/com/nytimes/android/external/fs/impl/FileSystemImpl.java
package com.nytimes.android.external.fs.impl;

import com.nytimes.android.external.cache.CacheLoader;
import com.nytimes.android.external.cache.LoadingCache;
import com.nytimes.android.external.fs.FileSystem;
import com.nytimes.android.external.fs.Util;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;

import okio.BufferedSource;

import static com.nytimes.android.external.cache.CacheBuilder.newBuilder;
import static java.lang.String.format;

/**
 * implements a {@link FileSystem} as regular files on disk in a specific document root (kind of like a root jail)
 * <p>
 * All operations are on the caller's thread.
 */
public class FileSystemImpl implements FileSystem {

    private final Util util = new Util();
    final LoadingCache<String, FSFile> files;
    final File root;

    public FileSystemImpl(final File root) throws IOException {
        this.root = root;

        this.files = newBuilder().maximumSize(20)
                .build(new CacheLoader<String, FSFile>() {
                    @Override
                    public FSFile load(String path) throws IOException {
                        return new FSFile(root, path);
                    }
                });

        util.createParentDirs(root);
    }

    @Override
    public BufferedSource read(String path) throws FileNotFoundException {
        return getFile(path).source();
    }

    @Override
    public void write(String path, BufferedSource source) throws IOException {
        getFile(path).write(source);
    }

    @Override
    public void delete(String path) throws IOException {
        getFile(path).delete();
    }

    @Override
    public Collection<String> list(String directory) throws FileNotFoundException {

        Collection<FSFile> foundFiles = findFiles(directory);
        Collection<String> names = new ArrayList<>(foundFiles.size());
        Iterator<FSFile> iterator = foundFiles.iterator();
        while ((iterator.hasNext())) {
            names.add(iterator.next().path());
        }
        return names;
    }

    @Override
    public void deleteAll(String directory) throws FileNotFoundException {

        Collection<FSFile> foundFiles = findFiles(directory);
        Iterator<FSFile> iterator = foundFiles.iterator();
        while ((iterator.hasNext())) {
            iterator.next().delete();
        }
    }

    @Override
    public boolean exists(String path) {
        return getFile(path).exists();
    }

    private FSFile getFile(String path) {
        return files.getUnchecked(cleanPath(path));
    }

    private String cleanPath(String dirty) {
        return util.simplifyPath(dirty);
    }

    private Collection<FSFile> findFiles(String path) throws FileNotFoundException {

        File searchRoot = new File(root, util.simplifyPath(path));
        if (searchRoot.exists() && searchRoot.isFile()) {
            throw new FileNotFoundException(format("expecting a directory at %s, instead found a file", path));
        }

        Collection<FSFile> foundFiles = new ArrayList<>();
        BreadthFirstFileTreeIterator iterator = new BreadthFirstFileTreeIterator(searchRoot);
        while (iterator.hasNext()) {
            File file = (File) iterator.next();
            foundFiles.add(files.getUnchecked(util.simplifyPath(file.getPath()
                    .replaceFirst(root.getPath(), ""))));
        }
        return foundFiles;
    }
}

File: filesystem/src/main/java/com/nytimes/android/external/fs3/filesystem/FileSystemImpl.kt
package com.nytimes.android.external.fs3.filesystem

import com.nytimes.android.external.cache3.CacheBuilder.newBuilder
import com.nytimes.android.external.cache3.CacheLoader
import com.nytimes.android.external.cache3.LoadingCache
import com.nytimes.android.external.fs3.Util
import com.nytimes.android.external.store3.base.RecordState
import okio.BufferedSource
import java.io.File
import java.io.FileNotFoundException
import java.io.IOException
import java.lang.String.format
import java.util.*
import java.util.concurrent.TimeUnit

/**
 * implements a [FileSystem] as regular files on disk in a specific document root (kind of like a root jail)
 *
 *
 * All operations are on the caller's thread.
 */
internal class FileSystemImpl(private val root: File) : FileSystem {

    private val files: LoadingCache<String, FSFile>

    init {
        this.files = newBuilder().maximumSize(20)
                .build(object : CacheLoader<String, FSFile>() {
                    override fun load(path: String): FSFile {
                        return FSFile(root, path)
                    }
                })

        Util.createParentDirs(root)
    }

    @Throws(FileNotFoundException::class)
    override fun read(path: String): BufferedSource {
        return getFile(path)!!.source()
    }

    @Throws(IOException::class)
    override fun write(path: String, source: BufferedSource) {
        getFile(path)!!.write(source)
    }

    @Throws(IOException::class)
    override fun delete(path: String) {
        getFile(path)!!.delete()
    }

    @Throws(FileNotFoundException::class)
    override fun list(path: String): Collection<String> {
        val foundFiles = findFiles(path)
        val names = ArrayList<String>(foundFiles.size)
        for (foundFile in foundFiles) {
            names.add(foundFile.path())
        }
        return names
    }

    @Throws(FileNotFoundException::class)
    override fun deleteAll(path: String) {
        val foundFiles = findFiles(path)
        for (foundFile in foundFiles) {
            foundFile.delete()
        }
    }

    override fun exists(file: String): Boolean {
        return getFile(file)!!.exists()
    }

    override fun getRecordState(expirationUnit: TimeUnit, expirationDuration: Long, path: String): RecordState {
        val file = getFile(path)
        if (!file!!.exists()) {
            return RecordState.MISSING
        }
        val now = System.currentTimeMillis()
        val cuttOffPoint = now - TimeUnit.MILLISECONDS.convert(expirationDuration, expirationUnit)
        return if (file.lastModified() < cuttOffPoint) {
            RecordState.STALE
        } else {
            RecordState.FRESH
        }
    }

    private fun getFile(path: String): FSFile? {
        return files.getUnchecked(cleanPath(path))
    }

    private fun cleanPath(dirty: String): String =
            Util.simplifyPath(dirty)

    @Throws(FileNotFoundException::class)
    private fun findFiles(path: String): Collection<FSFile> {
        val searchRoot = File(root, Util.simplifyPath(path))
        if (searchRoot.exists() && searchRoot.isFile) {
            throw FileNotFoundException(format("expecting a directory at %s, instead found a file", path))
        }

        val foundFiles = ArrayList<FSFile>()
        val iterator = BreadthFirstFileTreeIterator(searchRoot)
        while (iterator.hasNext()) {
            val file = iterator.next() as File?
            foundFiles.add(files.getUnchecked(Util.simplifyPath(file!!.path
                    .replaceFirst(root.path.toRegex(), "")))!!)
        }
        return foundFiles
    }
}
